File: apps/api/src/lib/voyage.test.ts
import { describe, it, expect } from "vitest";
import { buildEmbeddingRequest, batchTexts } from "./voyage";

describe("buildEmbeddingRequest", () => {
  it("builds body for documents", () => {
    const body = buildEmbeddingRequest(["a", "b"], "document");
    expect(body.input).toEqual(["a", "b"]);
    expect(body.input_type).toBe("document");
    expect(body.model).toBe("voyage-3");
  });

  it("builds body for query", () => {
    const body = buildEmbeddingRequest(["hỏi gì đó"], "query");
    expect(body.input_type).toBe("query");
  });
});

describe("batchTexts", () => {
  it("chia đúng batch khi vượt kích thước", () => {
    const batches = batchTexts(["a", "b", "c", "d", "e"], 2);
    expect(batches).toEqual([["a", "b"], ["c", "d"], ["e"]]);
  });

  it("một batch khi nhỏ hơn kích thước", () => {
    expect(batchTexts(["a", "b"], 96)).toEqual([["a", "b"]]);
  });

  it("mảng rỗng → không có batch", () => {
    expect(batchTexts([], 96)).toEqual([]);
  });

  it("ném lỗi khi batchSize < 1", () => {
    expect(() => batchTexts(["a"], 0)).toThrow();
  });
});

File: services/agent-go/internal/rag/voyage_test.go
package rag

import "testing"

func TestBuildEmbeddingRequest(t *testing.T) {
	r := buildEmbeddingRequest([]string{"a", "b"}, "document")
	if r.Model != "voyage-3" {
		t.Fatalf("model: %s", r.Model)
	}
	if r.InputType != "document" {
		t.Fatalf("input_type: %s", r.InputType)
	}
	if len(r.Input) != 2 {
		t.Fatalf("input len: %d", len(r.Input))
	}
}

func TestBatchTexts(t *testing.T) {
	got := batchTexts([]string{"a", "b", "c", "d", "e"}, 2)
	if len(got) != 3 {
		t.Fatalf("want 3 batches, got %d", len(got))
	}
	if len(got[2]) != 1 {
		t.Fatalf("last batch len: %d", len(got[2]))
	}
	if len(batchTexts(nil, 96)) != 0 {
		t.Fatal("empty input → 0 batches")
	}
	if len(batchTexts([]string{"a"}, 0)) != 1 {
		t.Fatal("size<1 → coi như 1")
	}
}
